modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java

```diff
@@ -112,12 +104,50 @@ public Map<String, StorageEngine> createStorageEngines(
     /**
      * Collects {@link DataStorageConfigurationSchema data storage schema} fields (with {@link Value}).
      *
+     * @param polymorphicSchemaExtensions {@link PolymorphicConfigInstance Polymorphic schema extensions} that contain extensions of {@link
+     * DataStorageConfigurationSchema}.
      * @return Mapping: {@link DataStorageModule#name Data storage name} -> filed name -> field type.
+     * @throws IllegalStateException If the {@link DataStorageConfigurationSchema data storage schemas} are not valid.
      */
-    public Map<String, Map<String, Class<?>>> collectSchemasFields() {
+    public Map<String, Map<String, Class<?>>> collectSchemasFields(Collection<Class<?>> polymorphicSchemaExtensions) {
+        Map<String, Class<? extends DataStorageConfigurationSchema>> schemas = polymorphicSchemaExtensions.stream()
+                .filter(DataStorageConfigurationSchema.class::isAssignableFrom)
+                .collect(toUnmodifiableMap(
+                        schemaCls -> schemaName((Class<? extends DataStorageConfigurationSchema>) schemaCls),
+                        schemaCls -> (Class<? extends DataStorageConfigurationSchema>) schemaCls
+                ));
+
+        Set<String> dataStorageNames = modules.stream()
```

Comment:
Ok, so:
modules is a Set that you retrieve by calling "this.modules = modules.values();", it's a proxy.
And after that you convert it back into the map with streams and all this magic. Why can't you store Map as a field?

Reply:
Fix it.